File: docker/worker-customcode-dotnet/harness/samples/BufferTool/BufferTool.cs
// Copyright (c) Honua. All rights reserved.
// Licensed under the Elastic License 2.0. See LICENSE in the project root.

using System.Text.Json;
using Honua.CustomCode.Sdk;
using NetTopologySuite.Geometries;
using NetTopologySuite.IO;

namespace Honua.CustomCode.Samples;

/// <summary>
/// Sample custom-code GP tool: buffer an input WKT geometry and write GeoJSON. This is
/// the .NET mirror of the Python sample <c>buffer_tool.py</c>. A real tool would pull
/// features via <see cref="GpContext.Client"/> and/or read staged
/// <see cref="GpContext.Inputs"/>; here we accept a WKT geometry + a buffer distance in
/// <see cref="GpContext.Params"/> and emit a buffered polygon as a GeoJSON artifact.
/// </summary>
/// <remarks>
/// Entrypoint spec for this tool: <c>"BufferTool::Honua.CustomCode.Samples.BufferTool"</c>.
/// Expected <c>params_json</c>:
/// <c>{"wkt": "POINT (1 1)", "distance": 0.5, "out_name": "buffered.geojson"}</c>.
/// </remarks>
public sealed class BufferTool : IGeoprocessingTool
{
    /// <inheritdoc />
    public Task<GpResult> ExecuteAsync(GpContext context, CancellationToken cancellationToken)
    {
        if (!context.Params.TryGetProperty("wkt", out var wktElement) ||
            wktElement.ValueKind != JsonValueKind.String ||
            wktElement.GetString() is not { Length: > 0 } wkt)
        {
            return Task.FromResult(GpResult.Failed("params.wkt is required (a WKT geometry)."));
        }

        var distance = context.Params.TryGetProperty("distance", out var d) && d.ValueKind == JsonValueKind.Number
            ? d.GetDouble()
            : 0.0;
        var outName = context.Params.TryGetProperty("out_name", out var n) && n.ValueKind == JsonValueKind.String
            ? n.GetString()!
            : "buffered.geojson";
        if (!ArtifactNames.IsSimpleFileName(outName))
        {
            return Task.FromResult(GpResult.Failed(
                $"params.out_name '{outName}' must be a simple file name (no path separators or '..')."));
        }

        context.Log.Info($"buffering '{wkt}' by {distance}");
        context.Progress.Report(10.0, "parsing input");

        Geometry geometry;
        try
        {
            geometry = new WKTReader().Read(wkt);
        }
        catch (Exception ex) when (ex is not OutOfMemoryException)
        {
            // Intentional catch-all: WKT parsing can throw a variety of NTS/format
            // exceptions and this tool boundary reports every failure as a GpResult
            // rather than crashing the harness process.
            return Task.FromResult(GpResult.Failed($"failed to parse WKT: {ex.Message}"));
        }

        cancellationToken.ThrowIfCancellationRequested();
        context.Progress.Report(50.0, "buffering");

        var buffered = geometry.Buffer(distance);

        var geometryJson = new GeoJsonWriter().Write(buffered);
        using var doc = JsonDocument.Parse(geometryJson);
        var feature = new
        {
            type = "Feature",
            geometry = doc.RootElement,
            properties = new { source_wkt = wkt, distance },
        };

        // False positive: outName was already validated by ArtifactNames.IsSimpleFileName
        // above, so it is never rooted/absolute.
        var outPath = Path.Join(context.WorkDirectory, outName);
        File.WriteAllText(outPath, JsonSerializer.Serialize(feature));
        context.Progress.Report(90.0, "writing output");

        context.Output.AddArtifact(outName, outPath);
        context.Log.Info($"wrote {outName} ({new FileInfo(outPath).Length} bytes)");
        context.Progress.Report(100.0, "done");

        return Task.FromResult(GpResult.Succeeded($"buffered geometry written to {outName}"));
    }
}

File: docker/worker-customcode-python/harness/samples/buffer_tool.py
"""Sample custom-code GP tool: buffer an input geometry and write GeoJSON.

This is a trivial transform that demonstrates the ``execute(context)`` contract.
A real tool would pull features via ``context.client`` and/or read staged
``context.inputs``. Here we accept a WKT point + a buffer distance in
``context.params`` and emit a buffered polygon as a GeoJSON artifact.

Entrypoint spec for this tool: ``buffer_tool:execute``.

Expected ``params_json``::

    {"wkt": "POINT (1 1)", "distance": 0.5, "out_name": "buffered.geojson"}
"""

from __future__ import annotations

import json

# These imports resolve from the harness package, which is installed in the
# image. A tool may import shapely/geopandas/etc. freely (pre-installed).
from honua_customcode_harness import GpContext, GpResult


def execute(context: GpContext) -> GpResult:
    params = context.params
    wkt = params.get("wkt")
    if not wkt:
        return GpResult.failed("params.wkt is required (a WKT geometry).")

    distance = float(params.get("distance", 0.0))
    out_name = str(params.get("out_name", "buffered.geojson"))

    context.log.info(f"buffering {wkt!r} by {distance}")
    context.progress.report(10.0, "parsing input")

    try:
        from shapely import wkt as shapely_wkt
        from shapely.geometry import mapping
    except ImportError:  # pragma: no cover - shapely is baked into the image
        return GpResult.failed("shapely is not available in this runtime.")

    geom = shapely_wkt.loads(wkt)
    context.cancellation.raise_if_cancelled()
    context.progress.report(50.0, "buffering")

    buffered = geom.buffer(distance)
    feature = {
        "type": "Feature",
        "geometry": mapping(buffered),
        "properties": {"source_wkt": wkt, "distance": distance},
    }

    out_path = context.workdir / out_name
    out_path.write_text(json.dumps(feature), encoding="utf-8")
    context.progress.report(90.0, "writing output")

    context.output.add_artifact(out_name, out_path)
    context.log.info(f"wrote {out_name} ({out_path.stat().st_size} bytes)")
    context.progress.report(100.0, "done")

    return GpResult.succeeded(f"buffered geometry written to {out_name}")
